src/Transports.AspNetCore/GraphQLHttpMiddleware.cs

```diff
@@ -108,225 +285,409 @@ public virtual async Task InvokeAsync(HttpContext context, RequestDelegate next)
-        if (bodyGQLBatchRequest == null)
-        {
-            var stopwatch = ValueStopwatch.StartNew();
-            await RequestExecutingAsync(gqlRequest);
-            var result = await ExecuteRequestAsync(gqlRequest, userContext, executer, context.RequestServices, cancellationToken);
-
-            await RequestExecutedAsync(new GraphQLRequestExecutionResult(gqlRequest, result, stopwatch.Elapsed));
+        var userContext = await BuildUserContextAsync(context);
+        var result = await ExecuteRequestAsync(context, gqlRequest, context.RequestServices, userContext);
+        var statusCode = Options.ValidationErrorsReturnBadRequest && !result.Executed
+            ? HttpStatusCode.BadRequest
+            : HttpStatusCode.OK;
+        await WriteJsonResponseAsync(context, statusCode, result);
+    }
 
-            await WriteResponseAsync(context.Response, _serializer, cancellationToken, result);
+    /// <summary>
+    /// Handles a batched GraphQL request.
+    /// </summary>
+    protected virtual async Task HandleBatchRequestAsync(
+        HttpContext context,
+        RequestDelegate next,
+        IList<GraphQLRequest> gqlRequests)
+    {
+        var userContext = await BuildUserContextAsync(context);
+        var results = new ExecutionResult[gqlRequests.Count];
+        if (gqlRequests.Count == 1)
+        {
+            results[0] = await ExecuteRequestAsync(context, gqlRequests[0], context.RequestServices, userContext);
         }
-        // Execute multiple graphql requests in one batch
         else
         {
-            var executionResults = new ExecutionResult[bodyGQLBatchRequest.Count];
-            for (int i = 0; i < bodyGQLBatchRequest.Count; ++i)
+            // Batched execution with multiple graphql requests
+            if (!Options.BatchedRequestsExecuteInParallel)
             {
-                var gqlRequestInBatch = bodyGQLBatchRequest[i];
-
-                var stopwatch = ValueStopwatch.StartNew();
-                await RequestExecutingAsync(gqlRequestInBatch, i);
-                var result = await ExecuteRequestAsync(gqlRequestInBatch, userContext, executer, context.RequestServices, cancellationToken);
-
-                await RequestExecutedAsync(new GraphQLRequestExecutionResult(gqlRequestInBatch, result, stopwatch.Elapsed, i));
-
-                executionResults[i] = result;
+                for (int i = 0; i < gqlRequests.Count; i++)
+                {
+                    results[i] = await ExecuteRequestAsync(context, gqlRequests[i], context.RequestServices, userContext);
+                }
+            }
+            else
+            {
+                var resultTasks = new Task<ExecutionResult>[gqlRequests.Count];
+                for (int i = 0; i < gqlRequests.Count; i++)
+                {
+                    resultTasks[i] = ExecuteScopedRequestAsync(context, gqlRequests[i], userContext);
+                }
+                await Task.WhenAll(resultTasks);
```

Comment:
This implementation is simple and straighforward. It's good. But consider switching from `public bool BatchedRequestsExecuteInParallel { get; set; } = true;`
to something like
`BatchedRequestsConcurrencyLimit { get; set; } = 5;`

Reply:
Sounds like a good feature to add.  To make it easier to understand the options, I would add the option not replace it.

Reply:
50/50 . `BatchedRequestsExecuteInParallel = false` equals to `BatchedRequestsConcurrencyLimit  = 1` . You may use single option and treat 1 as a special case or use both and just ignore second if the first is `false`. Whatever.

Reply:
#793